src/mlpack/bindings/python/CMakeLists.txt

```
@@ -251,52 +264,98 @@ endif ()
+        DEPENDS generate_pyx_${name}
+        WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/src/mlpack/bindings/python/
+        COMMENT "Building Cython binding ${name}.so...")
+
+    add_dependencies(python build_pyx_${name})
+    add_dependencies(build_pyx_${name} generate_pyx_${name})
+    add_dependencies(generate_pyx_${name} python_configured)
+
+    # Add the convenience import to __init__.py.  Note that this happens during
+    # configuration.
+    file(APPEND ${CMAKE_BINARY_DIR}/src/mlpack/bindings/python/mlpack/__init__.py
+        "from .${name} import ${name}\n")
+  endif ()
+endmacro()
+
+# this macro creates .so files for all methods in the given group.
+# if using add_python_wrapper(), then must group bindings using group_bindings()
+macro (add_python_wrapper group_name)
+
+  foreach (method ${${group_name}_VALID_METHODS})
+    add_python_binding("${group_name}_${method}")
+  endforeach()
+
+  if(EXISTS ${CMAKE_BINARY_DIR}/src/mlpack/bindings/python/${group_name}_main_includes.hpp)
+    file(REMOVE ${CMAKE_BINARY_DIR}/src/mlpack/bindings/python/${group_name}_main_includes.hpp)
+  endif()
+
+  foreach(loc ${${group_name}_METHOD_MAIN_FILES})
+    file(APPEND ${CMAKE_BINARY_DIR}/src/mlpack/bindings/python/${group_name}_main_includes.hpp
+        "#include <${loc}>\n")
+  endforeach()
+
+  file(APPEND ${CMAKE_BINARY_DIR}/src/mlpack/bindings/python/mlpack/__init__.py
+      "from .${group_name} import *\n")
+
+  add_custom_command(OUTPUT ${CMAKE_BINARY_DIR}/src/mlpack/bindings/python/generate_py_wrapper_${group_name}.cpp
+  COMMAND ${CMAKE_COMMAND}
+      -DGENERATE_CPP_IN=${CMAKE_SOURCE_DIR}/src/mlpack/bindings/python/generate_py_wrapper.cpp.in
+      -DGENERATE_CPP_OUT=${CMAKE_BINARY_DIR}/src/mlpack/bindings/python/generate_py_wrapper_${group_name}.cpp
+      -DGROUP_NAME=${group_name}
+      -DGROUP_VALID_METHODS="${${group_name}_VALID_METHODS}"
+      -DCATEGORY="${${group_name}_CATEGORY}"
+      -P ${CMAKE_SOURCE_DIR}/CMake/ConfigureFile.cmake
+  DEPENDS ${CMAKE_SOURCE_DIR}/src/mlpack/bindings/python/generate_py_wrapper.cpp.in
+          ${CMAKE_SOURCE_DIR}/CMake/ConfigureFile.cmake)
+
+  add_executable(generate_py_wrapper_${group_name}
+      ${CMAKE_BINARY_DIR}/src/mlpack/bindings/python/generate_py_wrapper_${group_name}.cpp
+      ${CMAKE_SOURCE_DIR}/src/mlpack/bindings/python/print_wrapper_py.hpp
+      ${CMAKE_SOURCE_DIR}/src/mlpack/bindings/python/print_wrapper_py.cpp)
+
+  target_link_libraries(generate_py_wrapper_${group_name} mlpack ${MLPACK_LIBRARIES})
+
+  add_custom_command(TARGET generate_py_wrapper_${group_name} POST_BUILD
       COMMAND ${CMAKE_COMMAND}
-          -DPROGRAM=$<TARGET_FILE:generate_pyx_${name}>
-          -DOUTPUT_FILE=${CMAKE_BINARY_DIR}/src/mlpack/bindings/python/mlpack/${name}.pyx
+          -DPROGRAM=$<TARGET_FILE:generate_py_wrapper_${group_name}>
+          -DOUTPUT_FILE=${CMAKE_BINARY_DIR}/src/mlpack/bindings/python/mlpack/${group_name}.py
           -P ${CMAKE_SOURCE_DIR}/CMake/RunProgram.cmake)
 
```

I think we need to add dependencies for the `python` target here:

```suggestion
    add_dependencies(python generate_py_wrapper_${name})
    add_dependencies(generate_py_wrapper_${group_name} python_configured)
```

I think that's right---I don't think there is a build step here, we just need to make `generate_py_wrapper_${group_name}` and run it.